File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer , HashSet<Character>> c = new HashMap<>(); 
        Map<Pair<Integer , Integer> , HashSet<Character>> b = new HashMap<>();
        for(int i = 0 ; i < 9 ; i++){
            HashSet<Character> r = new HashSet<>();
            int row = i / 3;
            for(int j = 0 ; j < 9 ; j++){
                char ch = board[i][j];
                int column = j / 3;
                if(ch == '.') continue;
                Pair<Integer , Integer> p = new Pair(row , column);
                if(!c.containsKey(j)) c.put(j , new HashSet<>());
                if(!b.containsKey(p)) b.put(p , new HashSet<>());
                if(r.contains(ch) || c.get(j).contains(ch) || b.get(p).contains(ch)) return false;
                r.add(ch);
                c.get(j).add(ch);
                b.get(p).add(ch);
            }
        }
        return true;
    }
}

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxs = [set() for _ in range(9)]
        for i in range(9):
            row = rows[i]
            b = 0 
            if i >= 3:
                b = 3
            if i >= 6:
                b = 6
            for j in range(9):
                num = board[i][j]
                if num == ".":
                    continue
                col = cols[j]
                box = boxs[b + j // 3]
                if num in row or num in col or num in box:
                    # print(num in row)
                    # print(num in col)
                    # print(num in box)
                    # print(num, i, j)
                    return False
                row.add(num)
                col.add(num)
                box.add(num)
        return True